content/articles/How-to-start-up-with-Socket-Programming-with-Java/index.md

```diff
@@ -0,0 +1,331 @@
+- [](#)
+- [](#)
+
+### Introduction
+Socket programming is a way of connecting two computers on a network to allow the exchange of information between the two machines that are running over `JRE`(Java Runtime Environment). The connection can be through **connection-oriented protocol** or **connectionless** but for our case, we will be using connection-oriented which is TCP/IP.
+
+TCP is a connection-oriented protocol, which means that before data can be transferred, two computers must first establish a connection.
+
+Connectionless -For this type of connection, UDP protocol (User Datagram Protocol)also a layer 2 protocol, is used to provide a good picture where you do not necessarily need every packet to arrive at its desired destination.
+
+We will be using Client\Server architecture to illustrate more on sockets. Client and server communicate by writing to and reading from socket connection thus for a connection to happen between the server and a client, a socket must be present. The two machines should know each other's network information.
+
+### What is a Socket?
+A socket is an **endpoint** of communication, acting as a connection between two computers on a network. It contains a socket number that the TCP layer can use to identify the application to which data will be transmitted. An endpoint is made up of a port number and an IP address.
+
+There are two types of sockets
+
+1. Server socket - which waits for a request from clients
+2. Client socket - can be used to send and receive a request from clients.
+
+The client must know two information about the server:
+1.IP address of Server and 
+2.Port number
+
+>Port number 0-1024 are reserved for administrative purposes(e,g 21 for FTP,23 for Telnet,25 for email and 80 for HTTP).
+
+This happens when a TCP connection is being used between two machines:-
+
+1. server creates a ServerSocket object, to indicate which port number our communication is to occur. If an I/O  error occurs, exception handling mechanisms are implemented; **Throws** an **I/OException** or a **try** **catch** block is used.
+2. The accept() function of the ServerSocket class is called by the server. This method waits for a client to connect to the server on a specific port before returning.
+3. A client builds an object in the socket, giving the name of the server and the connecting port number as the server awaits.
+4. The Socket Class builder will try to connect the client to the server and port number provided. The client now has a socket object that, if communication is established, may connect with the server.
+5. The accept() method on the server produces a reference to a new server socket that is connected to the client's socket.
+ 
+If the connection becomes a success, communication can therefore be established using **I/O streams**. Each socket has an input and output stream. The client's OutputStream connects to an InputStream of the server while the server's OutputStream connects to the client's InputStream.
```

Comment:
> If the connection becomes a success, communication can therefore be established using **I/O streams**. Each socket has an input and output stream. The client's OutputStream connects to an InputStream of the server while the server's OutputStream connects to the client's InputStream.

To improve grammar and flow I would change this to: 

"If the connection is successful, communication can be established using **I/O streams**. Each socket has an input and output stream. The client's OutputStream connects to the server's InputStream while the server's OutputStream connects to the client's InputStream."